### ISSUE TITLE
Support for i18n in templates

### ISSUE THREAD

**Issue by sazzer** [SergioBenitez/Rocket#1140]
(Apologies in advance if this is the wrong forum for this)

I'm trying to work out how to best support i18n in my templates, and struggling.

I know how to register functions that can be called from my Tera templates, but these are entirely based on global data or values provided by the template, whereas it would be best for the locale data to be provided automatically from the request.

What I can't work out how to do is to have functions that are called from the template, but that automatically have access to request level details.

In short, I'm trying to avoid this:
```
{{ t(key="message.key", locale=locale }}
```
And needing to provide the locale every time it's called.

Is there some standard way to achieve this?

**Comment by jebrosen** [comment]
I know for example that [www.rust-lang.org](https://github.com/rust-lang/www.rust-lang.org) does do i18n in templates with a custom tera helper without passing locale to each invocation of the helper, but I have not familiar with the implementation off hand.

> Is there some standard way to achieve this?

Not that I know of yet, but if someone were to propose an API or crate implementing something nice we could probably adopt such a feature in `rocket_contrib`'s template support.

**Comment by mjanda** [comment]
Personally, I'm using https://github.com/justinas/gettext and askama templates.

I have translator accessible from my top-level template pre-initialized with user's locale and just do {{ c.tr().gettext("Foo") }}.

On top of that, I have `__()` function, that produces StringToTranslate newtype that intentionally doesn't provide Display trait implementation, so it's impossible to print untranslated string to template. I use it so I can produce messages intended for user anywhere without needing to have translator available - while making sure I'll translate them once...and just once.

Also, I have `translate` function when I need to use other locale (user A triggers sending system message to user B who uses different locale etc.).

`xgettext` is configured to gather strings from all those functions.

Instead of gettext, you might have a look at https://docs.rs/fluent/0.7.1/fluent/ - it's probably more flexible, but I'm not sure about usability for your use case.

**Comment by Weasy666** [comment]
Would be a cool feature. Kazimuth created a crate that heavily depends on macros to use fluent with templates https://github.com/kazimuth/baked_fluent. Don't know how useful this would be for Rocket.

**Comment by cedeber** [comment]
You can check [fluent-templates](https://crates.io/crates/fluent-templates). It supports both handlebars and tera. So far I remember this is what the official rust website uses. It is based on https://www.projectfluent.org, a Mozilla research projects (probably not anymore :-/)

**Comment by cedeber** [comment]
Also found that, very useful to create a language guard: https://crates.io/crates/accept-language

**Comment by SergioBenitez** [comment]
I think i18n is extraordinarily important, but given that Rocket doesn't ship a templating engine or similar artifacts, I'd consider this request relatively out of scope for Rocket itself. Should this fact change, we should revisit this issue.
